registry/storage/shares.go

```diff
@@ -8,130 +8,155 @@ import (
+type SharesListFunc = func(filters ...func(share *types.SSVShare) bool) []*types.SSVShare
 
-// Shares is the interface for managing shares
+// Shares is the interface for managing shares.
 type Shares interface {
 	eth1.RegistryStore
 
-	SaveShare(logger *zap.Logger, share *types.SSVShare) error
-	SaveShareMany(logger *zap.Logger, shares []*types.SSVShare) error
-	GetShare(key []byte) (*types.SSVShare, bool, error)
-	GetAllShares(logger *zap.Logger) ([]*types.SSVShare, error)
-	GetFilteredShares(logger *zap.Logger, f func(share *types.SSVShare) bool) ([]*types.SSVShare, error)
-	DeleteShare(key []byte) error
+	// Get returns the share for the given public key, or nil if not found.
+	Get(pubKey []byte) *types.SSVShare
+
+	// List returns a list of shares, filtered by the given filters (if any).
+	List(filters ...func(*types.SSVShare) bool) []*types.SSVShare
+
+	// Save saves the given shares.
+	Save(logger *zap.Logger, shares ...*types.SSVShare) error
+
+	// Delete deletes the share for the given public key.
+	Delete(pubKey []byte) error
 }
 
 type sharesStorage struct {
 	db     basedb.IDb
-	lock   sync.RWMutex
 	prefix []byte
+	shares map[string]*types.SSVShare
+	mu     sync.RWMutex
 }
 
-// NewSharesStorage creates new share storage
-func NewSharesStorage(db basedb.IDb, prefix []byte) Shares {
-	return &sharesStorage{
+func NewSharesStorage(logger *zap.Logger, db basedb.IDb, prefix []byte) (Shares, error) {
+	storage := &sharesStorage{
+		shares: make(map[string]*types.SSVShare),
 		db:     db,
-		lock:   sync.RWMutex{},
 		prefix: prefix,
 	}
-}
-
-// SaveShare validator share to db
-func (s *sharesStorage) SaveShare(logger *zap.Logger, share *types.SSVShare) error {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	err := s.saveUnsafe(logger, share)
+	err := storage.load(logger)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	return nil
+	return storage, nil
 }
 
-func (s *sharesStorage) saveUnsafe(logger *zap.Logger, share *types.SSVShare) error {
-	value, err := share.Encode()
-	if err != nil {
-		logger.Error("failed to serialize share", zap.Error(err))
-		return err
-	}
-	return s.db.Set(s.prefix, buildShareKey(share.ValidatorPubKey), value)
+// load reads all shares from db.
+func (s *sharesStorage) load(logger *zap.Logger) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.db.GetAll(logger, append(s.prefix, sharesPrefix...), func(i int, obj basedb.Obj) error {
+		val := &types.SSVShare{}
+		if err := val.Decode(obj.Value); err != nil {
+			return fmt.Errorf("failed to deserialize share: %w", err)
+		}
+		s.shares[hex.EncodeToString(val.ValidatorPubKey)] = val
+		return nil
+	})
 }
 
-// GetShare by key
-func (s *sharesStorage) GetShare(key []byte) (*types.SSVShare, bool, error) {
-	s.lock.RLock()
-	defer s.lock.RUnlock()
+func (s *sharesStorage) Get(pubKey []byte) *types.SSVShare {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 
-	return s.getUnsafe(key)
+	return s.shares[hex.EncodeToString(pubKey)]
```

Comment:
so all the shares are always in the cache right?

Reply:
> so all the shares are always in the cache right?

yeah, NewSharesStorage calls load which reads all shares from storage to memory, and each save updates both memory and storage